File: go/parsers/subscribe.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Subscribe struct {
	gen *gen.Subscribe
}

func NewSubscribeFields(g *gen.Subscribe) messages.SubscribeFields {
	return &Subscribe{gen: g}
}

func (s *Subscribe) RequestID() uint64 {
	return s.gen.RequestID()
}

func (s *Subscribe) Options() map[string]any {
	return map[string]any{}
}

func (s *Subscribe) Topic() string {
	topic, _ := s.gen.Topic()
	return topic
}

func SubscribeToCapnproto(m *messages.Subscribe) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	subscribe, err := gen.NewRootSubscribe(seg)
	if err != nil {
		return nil, err
	}

	subscribe.SetRequestID(m.RequestID())
	if err := subscribe.SetTopic(m.Topic()); err != nil {
		return nil, err
	}

	data, err := msg.Marshal()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeSubscribe, data), nil
}

func CapnprotoToSubscribe(data []byte) (*messages.Subscribe, error) {
	msg, err := capnp.Unmarshal(data)
	if err != nil {
		return nil, err
	}

	subscribe, err := gen.ReadRootSubscribe(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewSubscribeWithFields(NewSubscribeFields(&subscribe)), nil
}

File: python/wamp_msgs_capnp/parsers/subscribe.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.subscribe import Subscribe, ISubscribeFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "subscribe.capnp")
subscribe_capnp = capnp.load(str(module_file))


class SubscribeFields(ISubscribeFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID

    @property
    def options(self) -> dict[str, Any]:
        options = {}
        if self._gen.match:
            options["match"] = self._gen.match

        return options

    @property
    def topic(self) -> str:
        return self._gen.topic


def subscribe_to_capnproto(s: Subscribe) -> bytes:
    subscribe = subscribe_capnp.Subscribe.new_message()
    subscribe.requestID = s.request_id
    subscribe.topic = s.topic

    match = s.options.get("match", None)
    if match is not None:
        subscribe.match = match

    packed_data = subscribe.to_bytes_packed()

    return helpers.prepend_header(Subscribe.TYPE, packed_data, b"")


def capnproto_to_subscribe(data: bytes) -> Subscribe:
    message_data, _ = helpers.extract_message(data)
    subscribe_obj = subscribe_capnp.Subscribe.from_bytes_packed(message_data)

    return Subscribe(SubscribeFields(subscribe_obj))
